ocaml/libs/clock/timer.ml

```diff
@@ -0,0 +1,53 @@
+type t = {start: Ptime.t; elapsed: Mtime_clock.counter; timeout: Mtime.Span.t}
+
+type remaining = Spare of Mtime.Span.t | Excess of Mtime.Span.t
+
+let is_shorter a ~than:b = Mtime.Span.compare a b < 0
+
+let is_longer a ~than:b = Mtime.Span.compare a b > 0
+
+let start ~timeout =
+  {start= Ptime_clock.now (); elapsed= Mtime_clock.counter (); timeout}
+
+let timeout {timeout; _} = timeout
+
+let elapsed t = Mtime_clock.count t.elapsed
+
+let remaining t =
+  let elapsed = Mtime_clock.count t.elapsed in
+  let difference = Mtime.Span.abs_diff t.timeout elapsed in
+  if Mtime.Span.compare t.timeout elapsed > 0 then
+    Spare difference
+  else
+    Excess difference
+
+let expired t = match remaining t with Spare _ -> false | Excess _ -> true
+
+let deadline_of t =
+  Mtime.Span.to_uint64_ns t.timeout
+  |> Int64.to_float
+  |> Ptime.Span.of_float_s
+  |> Option.get
+  |> Ptime.(Span.add Ptime.(to_span t.start))
+  |> Ptime.Span.to_float_s
+
+let shorten_by dur t =
+  let timeout =
+    if is_longer dur ~than:t.timeout then
+      Mtime.Span.zero
+    else
+      Mtime.Span.abs_diff dur t.timeout
+  in
+  {t with timeout}
+
+let extend_by dur t =
+  let timeout = Mtime.Span.add dur t.timeout in
+  {t with timeout}
+
+(* Conversion functions *)
+let span_to_s span =
+  Mtime.Span.to_uint64_ns span |> Int64.to_float |> fun ns -> ns /. 1e9
+
+let s_to_span s =
+  let micros_of = Float.to_int (s *. 1_000_000.) in
+  Mtime.Span.(micros_of * us)
```

Comment:
I don't see this function in mtime 2.0, probably best to multiply with 1e9 and use `of_float_ns`?